File: leetcode/1350-remove-sub-folders-from-the-filesystem/solution.cpp
class Trie {
    bool is_leaf;
    map<string, Trie> children;
public:
    Trie() {
        is_leaf = false;
    }
    bool insert(string::iterator begin, string::iterator end) {
        if(is_leaf) {
            return false;
        }
        auto i = find(begin, end, '/');
        if(i == end) {
            children[string(begin, end)].is_leaf = true;
            return true;
        } else {
            return children[string(begin, i)].insert(i+1, end);
        }
    }
};
class Solution {
public:
    vector<string> removeSubfolders(vector<string>& folder) {
        sort(folder.begin(), folder.end(), [](string& a, string& b) {
            return a.size() < b.size();
        });
        vector<string> ret;
        Trie root;
        for(auto s: folder) {
            if(root.insert(s.begin(), s.end())) {
                ret.push_back(s);
            }
        }
        return ret;
    }
};

File: leetcode/1350-remove-sub-folders-from-the-filesystem/solution.rs
use std::collections::HashMap;

#[derive(Default)]
struct Trie {
    children: HashMap<String, Trie>,
    is_leaf: bool,
}

impl Trie {
    pub fn insert(&mut self, value: &str) -> bool {
        if self.is_leaf {
            return false;
        }
        if let Some((value, rest)) = value.split_once('/') {
            return self.children
                .entry(value.to_string())
                .or_default()
                .insert(rest);
        } else {
            self.children
                .entry(value.to_string())
                .or_default()
                .is_leaf = true;
            return true;
        }
    }
}

impl Solution {
    pub fn remove_subfolders(mut folder: Vec<String>) -> Vec<String> {
        folder.sort_by(|a,b| a.len().cmp(&b.len()));
        let mut ret = Vec::new();
        let mut root = Trie::default();
        for s in folder {
            if root.insert(&s) {
                ret.push(s);
            }
        }
        return ret;
    }
}
